src/EVM/Expr.hs

```diff
@@ -880,6 +880,9 @@ simplify e = if (mapExpr go e == e)
       | b == c = a
       | otherwise = sub (add a b) c
 
+    -- Add is associative
+    go (Add (Add a b) c) = add a (add b c)
```

Comment:
I am wondering if there is a way to do this only if it is beneficial. 

For instance if the initial expression is `(Lit 17 + Lit 42) + x` then rearranging the expression will miss an opportunity for further simplification.

Reply:
Oh wow, good point! I think we should be canonical, so it should *eventually* always be done. But you are right that here we are very much missing an opportunity. What about this (for both `Add` and `Mul`):

```
    -- Add is associative
    go (Add (Add a b) c) = add a (add b c)
    go (Add (Lit a) (Add (Lit b) x) = add (Lit (a+b) x
```

Notice that the second `Add` will always have `Lit` as 1st argument thanks to canonicalization done by `add` (in case any of its arguments are a `Lit`). The above ensures that if *eventually* somehow, the second `Add`'s any argument becomes a `Lit` and the 1st `Add`'s 1st argument becomes a `Lit` then we simplify.

This should work for `Mul` as well. The only special case there is `Mul` by `Lit 0` and `Lit 1`, but that work with and 2nd argument, so the 2nd argument can be another `Mul`.

What do you think?

Reply:
Yes! I think adding this rule would be great!